File: AIN.Visitors.Mrz/Models/ParsedMrzResult.cs
using System;
using System.Collections.Generic;

namespace AIN.Visitors.Mrz.Models
{
    /// <summary>
    /// العقد الموحد المؤرخ لمخرجات تحليل MRZ حسب شروط المخطط المعتمد (القسم 3 و 4 و 10).
    /// </summary>
    public class ParsedMrzResult
    {
        public string ContractVersion { get; set; } = "1.0.0";
        public string ParserVersion { get; set; } = "2.0.0";

        public string DocumentFormat { get; set; } = string.Empty; // TD1, TD2, or TD3
        public string DocumentType { get; set; } = string.Empty; // e.g., P, I, A, C
        public string IssuingState { get; set; } = string.Empty; // رمز الدولة المصدرة (3 أحرف ICAO)
        public string Nationality { get; set; } = string.Empty; // رمز الجنسية (3 أحرف ICAO)

        public string DocumentNumber { get; set; } = string.Empty;
        public string Surname { get; set; } = string.Empty; // الاسم الأخير المطبّع بدون أحرف التعبئة <
        public string GivenNames { get; set; } = string.Empty; // الأسماء الأولى المطبّعة بدون أحرف التعبئة <
        public string DateOfBirth { get; set; } = string.Empty; // التنسيق القياسي المؤرخ ISO Format: YYYY-MM-DD
        public string? Sex { get; set; } = null; // M, F, X أو null في حال عدم التحديد
        public string DateOfExpiry { get; set; } = string.Empty; // التنسيق القياسي المؤرخ ISO Format: YYYY-MM-DD
        public string? OptionalData { get; set; } = null; // البيانات الاختيارية

        public Dictionary<string, bool> CheckDigitResults { get; set; } = new Dictionary<string, bool>();
        public bool CompositeCheckDigitResult { get; set; }
        public string ValidationStatus { get; set; } = "Unknown"; // حالات: Valid, Invalid, Warning
        public List<string> ErrorCodes { get; set; } = new List<string>(); // قائمة الرموز الدقيقة من فهرس الأخطاء

        public string ParsedAt { get; set; } = DateTime.UtcNow.ToString("o"); // التوقيت العالمي ISO 8601 UTC
        public string CorrelationId { get; set; } = Guid.NewGuid().ToString(); // معرّف التتبع الفريد
    }
}

File: ain-visitors-mrz-ts/ParsedMrzResult.ts
/**
 * العقد الموحد المؤرخ لمخرجات تحليل MRZ (نسخة TypeScript المطابقة).
 * متطابق هندسياً بنسبة 100% مع كلاس C# لضمان نجاح مقارنة مخرجات JSON الآلية (Canonical Parity).
 */

export interface CheckDigitResultsMap {
    documentNumber: boolean;
    dateOfBirth: boolean;
    dateOfExpiry: boolean;
    optionalData?: boolean;
}

export interface ParsedMrzResult {
    contractVersion: string;
    parserVersion: string;
    documentFormat: 'TD1' | 'TD2' | 'TD3' | '';
    documentType: string;
    issuingState: string;
    nationality: string;
    documentNumber: string;
    surname: string;
    givenNames: string;
    dateOfBirth: string; // ISO Format: YYYY-MM-DD
    sex: 'M' | 'F' | 'X' | null;
    dateOfExpiry: string; // ISO Format: YYYY-MM-DD
    optionalData: string | null;
    checkDigitResults: CheckDigitResultsMap;
    compositeCheckDigitResult: boolean;
    validationStatus: 'Valid' | 'Invalid' | 'Warning' | 'Unknown';
    errorCodes: string[];
    parsedAt: string; // ISO 8601 UTC
    correlationId: string;
}

/**
 * دالة مساعدة لإنشاء نتيجة معيارية أولية مع الحفاظ على الترتيب الصارم للحقول في مخرجات JSON.
 */
export function createCanonicalResult(): ParsedMrzResult {
    return {
        contractVersion: "1.0.0",
        parserVersion: "2.0.0",
        documentFormat: "",
        documentType: "",
        issuingState: "",
        nationality: "",
        documentNumber: "",
        surname: "",
        givenNames: "",
        dateOfBirth: "",
        sex: null,
        dateOfExpiry: "",
        optionalData: null,
        checkDigitResults: {
            documentNumber: false,
            dateOfBirth: false,
            dateOfExpiry: false
        },
        compositeCheckDigitResult: false,
        validationStatus: "Unknown",
        errorCodes: [],
        parsedAt: new Date().toISOString(),
        correlationId: "00000000-0000-0000-0000-000000000000"
    };
}